### DISCUSSION TITLE
Issue uploading attachments from web vault

### DISCUSSION THREAD

**Discussion by KnightTim** [dani-garcia/vaultwarden#2573] [Q&A] [answered]
See my comments here: https://github.com/dani-garcia/vaultwarden/issues/2516#issuecomment-1163894543 

### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.25.0
* Web-vault version: v2022.05.0
* Running within Docker: false (Base: Debian)
* Environment settings overridden: false
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.35.4

I get this error:
![image](https://user-images.githubusercontent.com/858797/175209492-5493e748-8c77-440d-b017-e320d7733b38.png)

In the logs I see: 
```
[request][INFO] POST /api/ciphers/a8e19efd-1fe9-42fa-852c-ea8902309576/attachment/v2
[response][INFO] (post_attachment_v2) POST /api/ciphers/<uuid>/attachment/v2 => 200 OK
[request][INFO] POST /api/ciphers/a8e19efd-1fe9-42fa-852c-ea8902309576/attachment/ba93d78c57068cd0f4d3
[response][INFO] (post_attachment_v2_data) POST /api/ciphers/<uuid>/attachment/<attachment_id> multipart/form-data => 400 Bad Request
[request][INFO] DELETE /api/ciphers/a8e19efd-1fe9-42fa-852c-ea8902309576/attachment/ba93d78c57068cd0f4d3
[response][INFO] (delete_attachment) DELETE /api/ciphers/<uuid>/attachment/<attachment_id> => 200 OK
[request][INFO] GET /api/ciphers/a8e19efd-1fe9-42fa-852c-ea8902309576
[response][INFO] (get_cipher) GET /api/ciphers/<uuid> => 200 OK
```

@BlackDex Suggested it might be an issue with my reverse proxy, I'm using nginx. 

Here is what my config has been for over a year. (It did work in the past for uploading attachments) 

```
server {
    listen 443 ssl http2;
    listen 8443 ssl http2;
    server_name my.domain;

    client_max_body_size 128M;

    ssl_certificate /etc/letsencrypt/live/my.domain/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/my.domain/privkey.pem;
    ssl_protocols TLSv1.2;
    ssl_prefer_server_ciphers on;
    ssl_ciphers "ECDH+AESGCM:EDH+AESGCM:AES256+EECDH:AES256+EDH";
    ssl_ecdh_curve secp384r1;
    ssl_session_cache shared:SSL:10m;
    ssl_session_tickets off;
    ssl_stapling on;
    ssl_stapling_verify on;
    resolver 127.0.0.1 valid=300s;
    resolver_timeout 5s;
    add_header X-Content-Type-Options nosniff;
    add_header Strict-Transport-Security "max-age=63072000; preload";
    keepalive_timeout 300s;

    location / {
        proxy_pass http://127.0.0.1:1443;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
    }

    location /notifications/hub {
        proxy_pass http://127.0.0.1:9455;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";
    }

    location /notifications/hub/negotiate {
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_pass http://127.0.0.1:1443;
    }
}
```

I'm not an expert on nginx so I tried the suggested config on the wiki [here](https://github.com/dani-garcia/vaultwarden/wiki/Proxy-examples) and it still gave me the same error. 

I have tried in both Firefox and MS Edge (from the same Windows box).  I can still download attachments. 

Let me know if there is any info I'm missing and thanks for any help.

**Comment by BlackDex** [comment]
One thing i notice is that you are using v1.25.0 with the web-vault v2022.05.0 which is strange since that version should have v2.28.1 as it's web-vault.

Also, do you have some Anti Virus tools installed, those sometimes cause an issue.

How large is the attachment? What happens if you just upload a simple text file with `hello` in it or something.

**Reply by KnightTim** [reply]
I can revert to an older version of the web-vault, I didn't realize there were issues running a newer web-vault with an older server. v1.25.0 is the latest version of the server. 

No AV software, I'm hosting it on a VPS running debian. (Bare metal, no docker either) 

I was testing with a small .txt file, only 1KB.

**Reply by KnightTim** [reply]
Just reverted to 2.28.1 and the same issue is present.

**Reply by BlackDex** [reply]
How did you installed Vaultwarden?

Well newer versions of the web-vault may need some server endpoint adjustments. While there are some bugs fixed with some new PR's nothing regarding attachments as far as i know.

Could you enable trace logging and see if there is something useful in the logs or post them somewhere?

**Reply by KnightTim** [reply]
I compile Vaultwarden from source, I primarily followed [this guide](https://pieterhollander.nl/post/bitwarden/)

Not sure how I would go about enabling trace logging, but if you point me to instruction I can give it a try.

**Reply by KnightTim** [reply]
Found the logging setting in the config file for Vaultwarden. I enabled the logging but it's getting late for me so I'll have to look through it tomorrow, it looks like it has some sensitive info so I don't want to upload it publicly.

**Reply by BlackDex** [reply]
@KnightTim Not sure if  you were able to do some checking yet, or that it is fixed now.
But since you build it your self, you could try to make a small adjustment to the code to get a bit more information.
try setting the following in the `main.rs` file at [Line 148](https://github.com/dani-garcia/vaultwarden/blob/main/src/main.rs#L148) to this:
```rust
        // Silence rocket logs
        .level_for("_", log::LevelFilter::Debug)
```
Or, for more details (but probably to much)
```rust
        // Silence rocket logs
        .level_for("_", log::LevelFilter::Trace)
```

That should give a bit more information on why it returns a 400.

**Reply by KnightTim** [reply]
I know it's been awhile, but I finally updated my vaultwarden install, and it's working from the web again. I believe from reading the issue thread, I would need to run the testing version for it to work from mobile. I'll just wait for that to get merged into stable, for now.

**Comment by BlackDex** [comment]
Maybe you can try what happens if you bypass nginx and look of without it you can upload a file, to try and rule items out.

**Reply by KnightTim** [reply]
Sorry, I am planning on trying your suggestions, just life has gotten a little busy and I haven't been able to check them

**Comment by 70599** [comment]
Same issue here running within Docker.

**Reply by BlackDex** [reply]
@70599 which version/type of the container, which reverse proxy, browser, OS?? What happens if you take the reverse proxy out of it, or what do it's logs say.

**Reply by 70599** [reply]
`docker version`
```shell
# docker version
docker version                                                                           0.000s
Client:
 Version:           20.10.16
 API version:       1.41
 Go version:        go1.17.10
 Git commit:        aa7e414fdcb23a66e8fabbef0a560ef1769eace5
 Built:             Sun May 15 15:07:52 2022
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server:
 Engine:
  Version:          20.10.16
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.17.10
  Git commit:       f756502055d2e36a84f2068e6620bea5ecf09058
  Built:            Sun May 15 14:59:12 2022
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.6.6
  GitCommit:        10c12954828e7c7c9b6e0ea9b0c02b01407d3ae1
 runc:
  Version:          1.1.2
  GitCommit:
 docker-init:
  Version:          0.19.0
  GitCommit:
```

valutwarden part of docker-compose.yaml
```yaml
  bitwarden:
    image: vaultwarden/server:alpine
    container_name: bitwarden
     volumes:
    - /etc/localtime:/etc/localtime:ro
    - $DATA/_app/bitwarden:/data
    - $DATA/log/bitwarden:/var/log/bitwarden
    environment:
      ROCKET_PORT: 60888
      WEBSOCKET_ENABLED: 'true'
      SIGNUPS_ALLOWED: 'false'
      LOG_FILE: /var/log/bitwarden/bitwarden.log
      LOG_LEVEL: WARN
    restart: unless-stopped
```

Reverse proxy with NGINX, basicly same as [wiki](https://github.com/dani-garcia/vaultwarden/wiki/Proxy-examples).
```conf
server {
	listen            unix:/dev/shm/nginx.sock       ssl http2;
	set_real_ip_from  unix:;
	server_name	      vault.example.com;

	include  global/ssl.example.com.conf;

	client_max_body_size  128M;

	location / {
		proxy_http_version  1.1;
		proxy_set_header    "Connection" "";

		proxy_set_header  Host $host;
		proxy_set_header  X-Real-IP $remote_addr;
		proxy_set_header  X-Forwarded-For $proxy_add_x_forwarded_for;
		proxy_set_header  X-Forwarded-Proto $scheme;
		
		proxy_pass  http://bitwarden:60888;
	}

	location /notifications/hub {
		proxy_http_version  1.1;

		proxy_set_header  Upgrade $http_upgrade;
		proxy_set_header  Connection "upgrade";
		proxy_set_header  Host $host;
		proxy_set_header  X-Real-IP $remote_addr;
		proxy_set_header  Forwarded $remote_addr;
		proxy_set_header  X-Forwarded-For $proxy_add_x_forwarded_for;
		proxy_set_header  X-Forwarded-Proto $scheme;

		proxy_pass  http://bitwarden:3012;
	}

	location /notifications/hub/negotiate {
		proxy_http_version  1.1;
		proxy_set_header    "Connection" "";

		proxy_set_header  Host $host;
		proxy_set_header  X-Real-IP $remote_addr;
		proxy_set_header  X-Forwarded-For $proxy_add_x_forwarded_for;
		proxy_set_header  X-Forwarded-Proto $scheme;

		proxy_pass  http://bitwarden:60888;
	}
}
```

All images are latest.

> What happens if you take the reverse proxy out of it

Not tested(Sorry not have time to debugging it).
Everything works except uploading attachments with this combination.

**Reply by 70599** [reply]
Switching vaultwarden image version to `testing-alpine` and then uploading attachments successed!
(Nothing else touched)

**Comment by KnightTim** [comment, marked as answer]
Overcome by updates to VaultWarden 

EDIT: Update for fix I had to make on 1.26.0.

 In case anyone else is running VaultWarden bare metal and has a similar setup for some reason the tmp folder under /data was set to user root and I had to change the owner to the the user I run the VaultWarden server under (bitwarden in my case) to be able to upload attachments from either mobile or the web.

**Comment by sergelogvinov** [comment]
I do not know, may be someone have the same issue...
I am using bitwarden cli to attach the file and i've got the same error.

part of the headers:

```
User-Agent: Bitwarden_CLI/2023.3.0 (MACOS)
Content-Type: multipart/form-data;boundary=--------------------------490709145970524404416406
Bitwarden-Client-Name: cli
Bitwarden-Client-Version: 2023.3.0
Accept: */*
Accept-Encoding: gzip,deflate

----------------------------490709145970524404416406
Content-Disposition: form-data; name="data"; filename="..."
Content-Type: application/octet-stream

----------------------------490709145970524404416406
Content-Disposition: form-data; name="data"; filename="..."
Content-Type: application/octet-stream

....binary data
```

Thought Browser all work fine. 
But it sends only one time `form-data`

**Reply by BlackDex** [reply]
Did you used `brew` to install the cli? In that case, the issue is that brew is using a newer nodejs/npm then what Bitwarden supports.
I think someone already created a bug report at brew regarding this, but I'm not sure.

Try to download the MacOS version here: https://github.com/bitwarden/clients/releases/tag/cli-v2023.3.0
Those are build by Bitwarden them self, and those should work just fine.

**Reply by sergelogvinov** [reply]
Oh, You definitely help me... Thank you!!!
